clusterloader2/pkg/measurement/common/service_creation_latency.go

```diff
@@ -0,0 +1,331 @@
+}
+
+func (s *serviceCreationLatencyMeasurement) handleObject(oldObj, newObj interface{}) {
+	var oldService *corev1.Service
+	var newService *corev1.Service
+	var ok bool
+	oldService, ok = oldObj.(*corev1.Service)
+	if oldObj != nil && !ok {
+		klog.Errorf("%s: uncastable old object: %v", s, oldObj)
+		return
+	}
+	newService, ok = newObj.(*corev1.Service)
+	if newObj != nil && !ok {
+		klog.Errorf("%s: uncastable new object: %v", s, newObj)
+		return
+	}
+	isEqual := oldService != nil &&
+		newService != nil &&
+		equality.Semantic.DeepEqual(oldService.Spec, newService.Spec) &&
+		equality.Semantic.DeepEqual(oldService.Status, newService.Status)
+	if isEqual {
+		return
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if !s.isRunning {
+		return
+	}
+	if newObj == nil {
+		if err := s.deleteObjectLocked(oldService); err != nil {
+			klog.Errorf("%s: delete checker error: %v", s, err)
+		}
+		return
+	}
+	if err := s.updateObjectLocked(newService); err != nil {
+		klog.Errorf("%s: create checker error: %v", s, err)
+	}
+}
+
+func (s *serviceCreationLatencyMeasurement) deleteObjectLocked(svc *corev1.Service) error {
+	key, err := runtimeobjects.CreateMetaNamespaceKey(svc)
+	if err != nil {
+		return fmt.Errorf("meta key created error: %v", err)
+	}
+	if pc, exists := s.pingCheckers[key]; exists {
+		pc.stop()
+		delete(s.pingCheckers, key)
+	}
+	return nil
+}
+
+func (s *serviceCreationLatencyMeasurement) updateObjectLocked(svc *corev1.Service) error {
+	if len(svc.Spec.Ports) < 1 {
+		return nil
+	}
+	key, err := runtimeobjects.CreateMetaNamespaceKey(svc)
+	if err != nil {
+		return fmt.Errorf("meta key created error: %v", err)
+	}
+	if _, exists := s.creationTimes.Get(key, creatingPhase); !exists {
+		s.creationTimes.Set(key, creatingPhase, time.Now())
+	}
+	if len(svc.Status.LoadBalancer.Ingress) < 1 {
+		return nil
+	}
+	if _, exists := s.creationTimes.Get(key, ipAssigningPhase); exists {
+		return nil
+	}
+	s.creationTimes.Set(key, ipAssigningPhase, time.Now())
+	klog.Infof("%s: %s ipAssigned", s, key)
+
+	s.pingCheckers[key] = &pingChecker{
+		owner:  s,
+		key:    key,
+		svc:    svc,
+		stopCh: make(chan struct{}),
+	}
+	go s.pingCheckers[key].run()
+
+	return nil
+}
+
+type pingChecker struct {
+	owner  *serviceCreationLatencyMeasurement
+	key    string
+	svc    *corev1.Service
+	stopCh chan struct{}
+}
+
+func (p *pingChecker) run() {
+	success := 0
+	for {
+		select {
+		case <-p.stopCh:
+			return
+		default:
+			if _, exists := p.owner.creationTimes.Get(p.key, reachabilityPhase); exists {
+				return
+			}
+			_, err := execservice.RunCommand(
+				fmt.Sprintf("curl %s:%d", p.svc.Status.LoadBalancer.Ingress[0].IP, p.svc.Spec.Ports[0].Port))
+			if err != nil {
+				success = 0
+				time.Sleep(pingBackoff)
+				continue
+			}
+			success++
+			if success == pingChecks {
+				func() {
+					klog.Infof("%v: %s is reachable", p.owner, p.key)
+					p.owner.lock.Lock()
+					defer p.owner.lock.Unlock()
```

Comment:
Please make measurementutil.ObjectTransitionTimes - this will simplify couple things in this file.

Reply:
Done.